src/rust/engine/fs/src/directory.rs

```diff
@@ -539,15 +622,29 @@ impl DigestTrie {
       ));
       f(&PathBuf::new(), &root);
     }
-    self.walk_helper(PathBuf::new(), f)
+    self.walk_helper(PathBuf::new(), symlink_behavior, f)
   }
 
-  fn walk_helper(&self, path_so_far: PathBuf, f: &mut impl FnMut(&Path, &Entry)) {
+  fn walk_helper(
+    &self,
+    path_so_far: PathBuf,
+    symlink_behavior: SymlinkBehavior,
+    f: &mut impl FnMut(&Path, &Entry),
+  ) {
     for entry in &*self.0 {
       let path = path_so_far.join(entry.name().as_ref());
-      f(&path, entry);
+
+      match (&symlink_behavior, entry) {
+        (SymlinkBehavior::Oblivious, Entry::Symlink(s)) => {
+          if let Ok(Some(entry)) = self.entry(&s.target) {
+            f(&path, entry);
+          };
```

Comment:
So, I _think_ that this needs to trigger recursion of `walk_helper`. Because if the destination is a directory, the effect of walking a symlink should actually be to continue walking within the destination, but with a new `path_so_far`.

Currently, it's only using the `entry` to call the user function, but it also needs to use that entry to continue recursing in `walk_helper` below.

To test that, you'd want to validate that:
```python
CreateDigest([
  FileEntry("a/1.txt", ..)
  SymlinkEntry("b", "a"),
])
```
... expands to the files:
```python
[
  "a/1.txt",
  "b/1.txt",
]
```